### PR TITLE
Refactoring integration tests

### PR DESCRIPTION
**Pull request by egorka-sereda**
refactor integration tests
remove F+1 test scenario

### CODE REVIEW

File: integration/qbft/tests/setup_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+func GetSharedData(t *testing.T) SharedData { //singleton B-)
+	if sharedData == nil {
+		t.Fatalf("shared data hadn't setuped, try to run test with -test.main flag")
+	}
+
+	return *sharedData
+}
+
+func TestMain(m *testing.M) {
+	_ = logging.SetLogLevelRegex("ssv/.*", "debug") // for debugging TODO: ssv/.* or ssv/*. ?
```
Comment (aaron-blox) [inline integration/qbft/tests/setup_test.go:41]: why is the returned error being ignored?
```suggestion
	_ = logging.SetLogLevelRegex("ssv/.*", "debug") // for debugging TODO: ssv/.* or ssv/*. ?
```

File: integration/qbft/tests/setup_test.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		logger.Fatal("error creating and start local net", zap.Error(err))
+		return
+	}
+
+	nodes := map[spectypes.OperatorID]network.P2PNetwork{}
+	for i := 0; i < len(ln.Nodes); i++ {
+		storage := newStores(logger)
+		ln.Nodes[i].RegisterHandlers(protocolp2p.WithHandler(
+			protocolp2p.LastDecidedProtocol,
+			handlers.LastDecidedHandler(logger.With(zap.String("who", fmt.Sprintf("decided-handler-%d", i+1))), storage, ln.Nodes[i]),
```
Comment (aaron-blox) [inline integration/qbft/tests/setup_test.go:59]: use named logger

File: integration/qbft/tests/setup_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		), protocolp2p.WithHandler(
+			protocolp2p.DecidedHistoryProtocol,
+			handlers.HistoryHandler(logger.With(zap.String("who", fmt.Sprintf("history-handler-%d", i+1))), storage, ln.Nodes[i], 25),
```
Comment (aaron-blox) [inline integration/qbft/tests/setup_test.go:62]: use named logger

File: integration/qbft/tests/utils.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+var (
+	KeySet4Committee  = spectestingutils.Testing4SharesSet()
+	KeySet7Committee  = spectestingutils.Testing7SharesSet()
+	KeySet10Committee = spectestingutils.Testing10SharesSet()
+	KeySet13Committee = spectestingutils.Testing13SharesSet()
+)
+
+type DutyProperties struct {
+	Slot  Slot
+	Idx   phase0.ValidatorIndex
```
Comment (aaron-blox) [inline integration/qbft/tests/utils.go:45]: please name this field better. is it the duty index, validator index, an ID that is for property x?

File: integration/qbft/tests/utils.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+type StoredInstanceProperties struct {
+	Height specqbft.Height
+}
+
+func newStores(logger *zap.Logger) *qbftstorage.QBFTStores {
+	db, err := storage.GetStorageFactory(basedb.Options{
+		Type:   "badger-memory",
+		Path:   "",
+		Logger: zap.L(),
```
Comment (aaron-blox) [inline integration/qbft/tests/utils.go:57]: why not set logger as the logger being passed into this func?

File: integration/qbft/tests/utils.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	case spectypes.BNRoleAttester:
+		testingDuty = spectestingutils.TestingAttesterDuty
+	case spectypes.BNRoleAggregator:
+		testingDuty = spectestingutils.TestingAggregatorDuty
+	case spectypes.BNRoleProposer:
+		testingDuty = spectestingutils.TestingProposerDuty
+	case spectypes.BNRoleSyncCommittee:
+		testingDuty = spectestingutils.TestingSyncCommitteeDuty
+	case spectypes.BNRoleSyncCommitteeContribution:
+		testingDuty = spectestingutils.TestingSyncCommitteeContributionDuty
```
Comment (aaron-blox) [inline integration/qbft/tests/utils.go:90]: please add a default case which should probably return an error

File: integration/qbft/tests/utils.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+		SharePubKey:     keySet.Shares[id].GetPublicKey().Serialize(),
+		DomainType:      spectypes.PrimusTestnet,
+		Quorum:          keySet.Threshold,
+		PartialQuorum:   keySet.PartialThreshold,
+		Committee:       keySet.Committee(),
+	}
+}
+
+func getKeySet(committee int) *spectestingutils.TestKeySet {
+	switch (committee - 1) / 3 {
```
Comment (aaron-blox) [inline integration/qbft/tests/utils.go:159]: please add a comment about what is equation is

File: network/p2p/test_utils.go [hunk 8]
```diff
@@ -69,44 +66,59 @@ func (ln *LocalNet) WithBootnode(ctx context.Context, logger *zap.Logger) error
...
+		if err != nil {
+			return nil, err
+		}
+
+		eg, ctx := errgroup.WithContext(pCtx)
+		for i, node := range ln.Nodes {
+			i, node := i, node //hack to avoid closures. price of using error groups
+
+			eg.Go(func() error { //if replace EG to regular goroutines round don't change to second in test
+				if err := node.Start(); err != nil {
```
Comment (aaron-blox) [inline network/p2p/test_utils.go:83]: is node.Start() safe to call multiple times?

File: network/p2p/test_utils.go [hunk 9]
```diff
@@ -127,18 +141,18 @@ func (ln *LocalNet) NewTestP2pNetwork(ctx context.Context, keys testing.NodeKeys
...
 		}
 	}
 	i := 0
 	nodes, keys, err := testing.NewLocalNetwork(ctx, n, func(pctx context.Context, keys testing.NodeKeys) network.P2PNetwork {
 		i++
-		logger := loggerFactory(fmt.Sprintf("node-%d", i))
+		logger := logger.With(zap.String("who", fmt.Sprintf("node-%d", i)))
 		p, err := ln.NewTestP2pNetwork(pctx, keys, logger, forkVersion, n)
 		if err != nil {
 			logger.Error("could not setup network", zap.Error(err))
```
Comment (aaron-blox) [inline network/p2p/test_utils.go:158]: is the `<-time.After(time.Millisecond * 500)` on line 168 needed still?

**Review by aaron-blox — COMMENTED** [review]
some high level comments

File: network/p2p/test_utils.go (hunk 8 above)
Comment (egorka-sereda) [reply network/p2p/test_utils.go:83]: yeah, verified with Nikita and Amir

File: integration/qbft/tests/setup_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+func GetSharedData(t *testing.T) SharedData { //singleton B-)
+	if sharedData == nil {
+		t.Fatalf("shared data hadn't setuped, try to run test with -test.main flag")
+	}
+
+	return *sharedData
+}
+
+func TestMain(m *testing.M) {
+	if err := logging.SetLogLevelRegex("ssv/.*", "debug"); err != nil { // for debugging TODO: ssv/.* or ssv/*. ?
```
Comment (aaron-blox) [inline integration/qbft/tests/setup_test.go:41]:
```suggestion
	if err := logging.SetLogLevelRegex("ssv/.*", "debug"); err != nil {
```

**Review by aaron-blox — COMMENTED** [review]
please fix linting errors

File: integration/qbft/tests/scenario_test.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+		}
+
+		//invoking duties
+		for id, dutyProp := range s.Duties {
+			go func(id spectypes.OperatorID, dutyProp DutyProperties) { //launching goroutine for every validator
+				time.Sleep(dutyProp.Delay)
+
+				duty := createDuty(getKeySet(s.Committee).ValidatorPK.Serialize(), dutyProp.Slot, dutyProp.ValidatorIndex, role)
+				require.NoError(t, s.validators[id].StartDuty(duty))
+			}(id, dutyProp)
```
Comment (nivBlox) [inline integration/qbft/tests/scenario_test.go:66]: why not add to queue? in order to mock real behaviour

File: integration/qbft/tests/scenario_test.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+					break
+				}
+
+				time.Sleep(500 * time.Millisecond) // waiting for duty will be done and storedInstance would be saved
+			}
+
+			// logging stored instance
+			jsonInstance, err := json.Marshal(storedInstance)
+			require.NoError(t, err)
+			fmt.Println(string(jsonInstance))
```
Comment (nivBlox) [inline integration/qbft/tests/scenario_test.go:89]: need to print it?

File: integration/qbft/tests/scenario_test.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+	db, err := storage.GetStorageFactory(basedb.Options{
+		Type:   "badger-memory",
+		Path:   "",
+		Logger: logger,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	store := qbftstorage.New(db, logger, "integration-tests", protocolforks.GenesisForkVersion)
```
Comment (nivBlox) [inline integration/qbft/tests/scenario_test.go:144]: the prefix should be based on the role type  and not `"integration-tests"`
so need to have different instance for each role type

**Review by nivBlox — CHANGES_REQUESTED** [review]

File: integration/qbft/tests/scenario_test.go (hunk 12 above)
Comment (egorka-sereda) [reply integration/qbft/tests/scenario_test.go:89]: yeah, for debugging purposes

File: integration/qbft/tests/scenario_test.go (hunk 11 above)
Comment (lior-blox) [reply integration/qbft/tests/scenario_test.go:85]: @egorka-sereda @aaron-blox this is resolved?

File: integration/qbft/tests/regular_test.go [hunk 14]
```diff
@@ -0,0 +1,111 @@
+package tests
```
Comment (lior-blox) [inline integration/qbft/tests/regular_test.go:1]: @aaron-blox while this is still WIP, lets remove from the github actions to run integration tests until they are stable.
you can create a regex rule to run it if the branch starts/ends with integration-test

**Review by lior-blox — CHANGES_REQUESTED** [review]

File: integration/qbft/tests/scenario_test.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+		// teardown
+		for _, val := range s.validators {
+			require.NoError(t, val.Stop())
+		}
+	})
+}
+
+func getKeySet(committee int) *spectestingutils.TestKeySet {
+	switch committee {
```
[the remark this answers is not in the document]
Comment (egorka-sereda) [reply integration/qbft/tests/scenario_test.go]: sometimes we need different number of participants than 4, 7, 10, 13
for example in f+1 decided we want to launch only 3 validators, but we need keysets for them as for 4 validators-committee

**Comment by egorka-sereda** [comment]
fixed all comments
@lior-blox @aaron-blox 
waiting for next review round
Thanks!

File: integration/qbft/tests/scenario_test.go [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+
+		// teardown
+		for _, val := range s.validators {
+			require.NoError(t, val.Stop())
+		}
+	})
+}
+
+func getKeySet(committee int) *spectestingutils.TestKeySet {
+	switch committee { //sometimes we need different number of participants than 4, 7, 10, 13
```
Comment (aaron-blox) [inline integration/qbft/tests/scenario_test.go:116]:
```suggestion
// getKeySet returns the keyset for a given committee size. Some tests have a
// committee size smaller than 3f+1 in order to simulate cases where operators are offline
func getKeySet(committee int) *spectestingutils.TestKeySet {
	switch committee { 
```

File: integration/qbft/tests/scenario_test.go (hunk 11 above)
Comment (aaron-blox) [reply integration/qbft/tests/scenario_test.go:84]: is resolved

File: integration/qbft/tests/scenario_test.go (hunk 13 above)
Comment (aaron-blox) [reply integration/qbft/tests/scenario_test.go]: resolved

File: integration/qbft/tests/scenario_test.go [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+		SharePubKey:     keySet.Shares[id].GetPublicKey().Serialize(),
+		DomainType:      spectypes.PrimusTestnet,
+		Quorum:          keySet.Threshold,
+		PartialQuorum:   keySet.PartialThreshold,
+		Committee:       keySet.Committee(),
+	}
+}
+
+func quorum(committee int) int {
+	return (committee*2 + 1) / 3 // committee = 3f+1; quorum = 2f+1 // https://drive.google.com/file/d/1bP_MLq0MM7ZBSR0Ddh7HUPcc42vVUKwz/view?usp=share_link
```
[the remark this answers is not in the document]
Comment (aaron-blox) [reply integration/qbft/tests/scenario_test.go]: resolved

File: integration/qbft/tests/regular_test.go (hunk 14 above)
Comment (aaron-blox) [reply integration/qbft/tests/regular_test.go:1]: resolved

**Review by aaron-blox — APPROVED** [review]
i added a suggestion to change the comment but LGTM otherwise

File: integration/qbft/tests/round_change_test.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+	return func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
+		require.Equal(t, int(qbft.FirstHeight), int(actual.State.Height), "height not matching") //int conversion needs to show correct output from require
+		require.Equal(t, int(qbft.Round(2)), int(actual.State.Round), "round not matching")
+
+		require.NotNil(t, actual.DecidedMessage, "no decided message")
+		if quorum(committee) > len(actual.DecidedMessage.Signers) {
+			require.Fail(t, "no commit qourum")
+		}
+
+		require.Equal(t, int(spectypes.OperatorID(2)), int(actual.State.ProposeContainer.Msgs[2][0].Signers[0])) //on second round proposer is 2,
```
Comment (aaron-blox) [inline integration/qbft/tests/round_change_test.go:46]: you should add checks before doing slice access for everything here
`actual.State.ProposeContainer.Msgs[2][0].Signers[0]`
Each of these can cause a panic if you are trying to access something that does not exist

**Review by aaron-blox — APPROVED** [review]

File: .github/workflows/integration-test.yml [hunk 19]
```diff
@@ -1,14 +1,14 @@
...
 on:
-  push:
-    branches:
-      - "main"
-      - "stage"
-  pull_request:
-    branches:
-      - "main"
-      - "stage"
+#  push:
```
Comment (nivBlox) [inline .github/workflows/integration-test.yml:4]: if no need, let's remove

File: integration/qbft/tests/regular_test.go [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+	regular.Run(t, spectypes.BNRoleSyncCommitteeContribution)
+}
+
+func regularValidator() func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
+	return func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
+		require.Equal(t, int(qbft.FirstHeight), int(actual.State.Height), "height not matching") //int conversion needs to show correct output from require
+		require.Equal(t, int(qbft.FirstRound), int(actual.State.Round), "round not matching")
+
+		require.NotNil(t, actual.DecidedMessage, "no decided message")
+		if quorum(committee) > len(actual.DecidedMessage.Signers) {
```
Comment (nivBlox) [inline integration/qbft/tests/regular_test.go:107]: you can use require.Greater() instead

File: integration/qbft/tests/round_change_test.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+	//roundChange.Run(t, spectypes.BNRoleSyncCommitteeContribution) todo implement sync committee contribution role support
+}
+
+func roundChangeValidator() func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
+	return func(t *testing.T, committee int, actual *protocolstorage.StoredInstance) {
+		require.Equal(t, int(qbft.FirstHeight), int(actual.State.Height), "height not matching") //int conversion needs to show correct output from require
+		require.Equal(t, int(qbft.Round(2)), int(actual.State.Round), "round not matching")
+
+		require.NotNil(t, actual.DecidedMessage, "no decided message")
+		if quorum(committee) > len(actual.DecidedMessage.Signers) {
```
Comment (nivBlox) [inline integration/qbft/tests/round_change_test.go:42]: you can use require.Greater() instead

File: integration/qbft/tests/scenario_test.go [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+	db, err := storage.GetStorageFactory(basedb.Options{
+		Type:   "badger-memory",
+		Path:   "",
+		Logger: logger,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	store := qbftstorage.New(db, logger, role.String(), protocolforks.GenesisForkVersion)
```
Comment (nivBlox) [inline integration/qbft/tests/scenario_test.go:159]: need to create store for each role.
reference - operator/node.go:91

File: integration/qbft/tests/setup_test.go [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+var sharedData *SharedData
+
+type SharedData struct {
+	Logger *zap.Logger
+	Nodes  map[spectypes.OperatorID]network.P2PNetwork
+}
+
+func GetSharedData(t *testing.T) SharedData { //singleton B-)
+	if sharedData == nil {
```
Comment (nivBlox) [inline integration/qbft/tests/setup_test.go:30]: let's use require.NotNil

File: protocol/v2/qbft/controller/future_msg.go [hunk 24]
```diff
@@ -17,6 +17,7 @@ func (c *Controller) UponFutureMsg(msg *specqbft.SignedMessage) (*specqbft.Signe
 	}
 	if c.f1SyncTrigger() {
 		c.logger.Debug("triggered f+1 sync",
+			zap.Any("operator-id", c.Share.OperatorID),
```
Comment (nivBlox) [inline protocol/v2/qbft/controller/future_msg.go:20]: it's not a place for this log. please revert. we can open a issue to add this if needed

File: protocol/v2/qbft/instance/instance.go [hunk 25]
```diff
@@ -52,7 +52,7 @@ func NewInstance(
 		config:      config,
 		processMsgF: spectypes.NewThreadSafeF(),
 		metrics:     newMetrics(msgId),
-		logger: logger.With(zap.String("publicKey", hex.EncodeToString(msgId.GetPubKey())), zap.String("role", msgId.GetRoleType().String()),
+		logger: logger.With(zap.Int("operator-id", int(share.OperatorID)), zap.String("publicKey", hex.EncodeToString(msgId.GetPubKey())), zap.String("role", msgId.GetRoleType().String()),
```
Comment (nivBlox) [inline protocol/v2/qbft/instance/instance.go:55]: please revert it as it not needed in this scope

File: protocol/v2/ssv/validator/msgqueue_consumer.go [hunk 26]
```diff
@@ -102,7 +102,7 @@ func (v *Validator) ConsumeQueue(msgID spectypes.MessageID, handler MessageHandl
 
 		// Handle the message.
 		if err := handler(msg); err != nil {
-			v.logMsg(msg, "could not handle message", zap.Any("type", msg.SSVMessage.MsgType), zap.Error(err))
+			v.logMsg(msg, "could not handle message", zap.Int("operator-id", int(v.Share.OperatorID)), zap.Any("type", msg.SSVMessage.MsgType), zap.Error(err))
```
Comment (nivBlox) [inline protocol/v2/ssv/validator/msgqueue_consumer.go:105]: let's remove

File: protocol/v2/ssv/validator/startup.go [hunk 27]
```diff
@@ -20,6 +20,7 @@ func (v *Validator) Start() error {
 		for role, r := range v.DutyRunners {
 			share := r.GetBaseRunner().Share
 			if share == nil { // TODO: handle missing share?
+				v.logger.Warn("share is missing")
```
Comment (nivBlox) [inline protocol/v2/ssv/validator/startup.go:23]: let's add the runner role type to the log

**Review by nivBlox — CHANGES_REQUESTED** [review]

**Review by nivBlox — APPROVED** [review]

**Review by lior-blox — APPROVED** [review]

**Pull request merged**